reps/2023-07-20-torch-trainer-apis.md

```diff
@@ -0,0 +1,460 @@
+    ...
+    return transformers.Trainer(**trainer_kwargs)
+
+trainer = TransformersTrainer(
+    trainer_init_per_worker=trainer_init_per_worker,
+    trainer_init_config=trainer_init_config,
+    ...
+)
+trainer.fit()
+```
+
+#### `TorchTrainer` (Proposed)
+
+In this proposal, we provide a `prepare_trainer` utility that the user can use directly in their training loop code to validate and/or make modifications so that the `Trainer` object is configured correctly to be compatible with Ray Train.
+
+With this, the user is able define their training logic to use additional functionality such as `transformers.Trainer.evaluate()`.
+
+```python
+import transformers
+import ray.train.huggingface.transformers
+from ray.train.torch import TorchTrainer
+
+def train_func(config):
+    ...
+    transformers_trainer = transformers.Trainer(**trainer_kwargs)
+    transformers_trainer = ray.train.huggingface.transformers.prepare_trainer(transformers_trainer)
+    transformers_trainer.train()
+
+trainer = TorchTrainer(
+    train_loop_per_worker=train_func,
+    ...
+)
+trainer.fit()
+```
+
+**Reporting & Checkpointing:**
+
+For Transformers, we introduce a `RayTrainReportCallback`(`TrainerCallback`) that will report metrics and checkpoints to Ray Train.
+
+```python
+from transformers.trainer_callback import TrainerCallback
+
+class RayTrainReportCallback(TrainerCallback):
+    # Define logic for calling `ray.train.report` as a Callback.
+    ...
+
+def train_func():
+    ...
+    trainer = Trainer(...)
+    # Create checkpoint.
+    report_checkpoint_callback = RayTrainReportCallback()
+    trainer.add_callback(report_checkpoint_callback)
+    # Restore checkpoint.
+    checkpoint_path = None
+    checkpoint = ray.train.get_checkpoint()
+    if checkpoint:
+        checkpoint_dir = checkpoint.to_directory()
+        checkpoint_path = Path(checkpoint_dir) / CHECKPOINT_FILE_NAME
+    trainer.train(resume_from_checkpoint=checkpoint_path)
+```
+
+**Data Ingestion:**
+
+Transformers expects a `torch.util.data.Dataset` so we can directly create a `RayDataIterableDataset`.
+
+```python
+from torch.utils.data import IterableDataset
+
+class RayDataIterableDataset(IterableDataset):
+    def __init__(self, dataset):
+        self.dataset = dataset
+
+    def __iter__(self):
+        return self.dataset.iter_rows()
+
+def train_func():
+    train_dataset_shard = ray.train.get_dataset_shard("train")
+    train_dataset = RayDataIterableDataset(train_dataset_shard)
+    eval_dataset = ray.train.get_dataset_shard("eval")
+    eval_dataset = RayDataIterableDataset(val_dataset)
+    transformers_trainer = transformers.Trainer(..., train_dataset=train_dataset, eval_dataset=eval_dataset)
+    transformers_trainer.train()
+
+train_dataset = ray.data.read_parquet(...).map_batches(...)
+eval_dataset = ray.data.read_parquet(...).map_batches(...)
+trainer = ray.train.torch.TorchTrainer(train_func, datasets={"train": train_dataset, "eval": eval_dataset})
```

Comment:
Currently, we require the use of `DataConfig` for correctly validation data sharding, e.g.:
```py
trainer = ray.train.torch.TorchTrainer(
   train_func, 
   datasets={"train": train_dataset, "eval": eval_dataset},
   data_config=DataConfig(datasets_to_split=["train", "eval"]) # required
)
```
@woshiyyya I agree it would be good and result in fewer bugs if users could easily perform dataset sharding without the need for `data_config`.

Reply:
Hi @YiranJing , I think we still need to have `data_config`, but we can slightly change the default behavior from sharding "train" only to sharding all datasets.

So for most DDP training workloads, users don't have to provide an extra DataConfig. 

If some users don't want to shard specific datasets, they can then provide a `DataConfig`. e.g.

```python
trainer = ray.train.torch.TorchTrainer(
   train_func, 
   datasets={"train": train_dataset, "eval": eval_dataset, "special": no_shard_dataset}
   data_config=DataConfig(datasets_no_split=["special"])
)
```

Reply:
@YiranJing could you share more about your use case and desired behavior in https://github.com/ray-project/ray/issues/37668?

Reply:
replied here https://github.com/ray-project/ray/issues/37668#issuecomment-1652842343